### PR TITLE
Delocalize DE results to study bucket (SCP-4300)

### PR DESCRIPTION
**Pull request by jlchang**
Leverage ingest pipeline infrastructure to delocalize DE output files to the study bucket.
 
To test, run DE using files from an GCP bucket associated with Staging SCP300
Before you start, check the study bucket for delocalized output
`https://console.cloud.google.com/storage/browser/fc-9b0f3acd-cf8f-4f01-b736-0a5a7b94e8c9`
If there is an "_scp_internal" directory, delete it before you start.

activate the scp-ingest-pipeline repo virtualenv
then from the scripts directory of the scp-ingest-pipeline repo, perform this setup:

```
source ../scripts/setup_mongo_dev.sh <path to your Github token file>
unset BARD_HOST_URL
```

for the manual testing block, you will need to run the following if you get an authentication error from gcloud (Thanks Jon!):

`gcloud auth application-default login --no-launch-browser`

Run the DE job from the ingest directory of the scp-ingest-pipeline repo:
`python ingest_pipeline.py --study-id addedfeed000000000000000 --study-file-id dec0dedfeed1111111111111 differential_expression --annotation-name cell_type__ontology_label --annotation-type group --annotation-scope study --matrix-file-path gs://fc-9b0f3acd-cf8f-4f01-b736-0a5a7b94e8c9/de_integration.tsv --matrix-file-type dense --annotation-file gs://fc-9b0f3acd-cf8f-4f01-b736-0a5a7b94e8c9/de_integration_unordered_metadata.tsv --cluster-file gs://fc-9b0f3acd-cf8f-4f01-b736-0a5a7b94e8c9/de_integration_cluster.tsv --cluster-name de_integration --study-accession SCPde --differential-expression`

NOTE: I've renamed input files per PR comments but did not change the files that are currently in staging SCP300. I have updated the delocalization URL below to reflect the newly agreed upon DE file path URL (underscore instead of hyphen).

confirm that the job runs successfully and the following output are delocalized to
https://console.cloud.google.com/storage/browser/fc-9b0f3acd-cf8f-4f01-b736-0a5a7b94e8c9/_scp_internal/differential_expression

Note the full cloud path to the files includes "_scp_internal" and "differential_expression" folders 

> de_integration--cell_type__ontology_label--cholinergic_neuron--study--wilcoxon.tsv
> de_integration--cell_type__ontology_label--cranial_somatomotor_neuron--study--wilcoxon.tsv
> de_integration--cell_type__ontology_label--pyramidal_neuron--study--wilcoxon.tsv
> de_integration--cell_type__ontology_label--somatomotor_neuron--study--wilcoxon.tsv
> de_integration--cell_type__ontology_label--sympathetic_cholinergic_neuron--study--wilcoxon.tsv

### CODE REVIEW

File: ingest/ingest_pipeline.py [hunk 1]
```diff
@@ -34,12 +34,20 @@
 
 # Ingest mtx files
 python ingest_pipeline.py --study-id 5d276a50421aa9117c982845 --study-file-id 5dd5ae25421aa910a723a337 ingest_expression --taxon-name 'Homo sapiens' --taxon-common-name human --matrix-file ../tests/data/mtx/matrix.mtx --matrix-file-type mtx --gene-file ../tests/data/genes.tsv --barcode-file ../tests/data/barcodes.tsv
+
+# Differential Expression analysis (dense matrix)
+python ingest_pipeline.py --study-id addedfeed000000000000000 --study-file-id dec0dedfeed1111111111111 differential_expression --annotation-name cell_type__ontology_label --annotation-type group --annotation-scope study --matrix-file-path ../tests/data/differential_expression/de_integration.tsv --matrix-file-type dense --annotation-file ../tests/data/differential_expression/de_integration_unordered_metadata.tsv --cluster-file ../tests/data/differential_expression/de_integration_cluster.tsv --cluster-name de_integration --study-accession SCPdev --differential-expression
```
Comment (eweitz) [inline ingest/ingest_pipeline.py:39]: Nit: having a file named `de_integration.tsv` is a bit opaque.  Like `de_integration_unordered_metadata.tsv`, it'd be clearer to name the file something like `de_integration_dense_matrix`.  

(I could also see trimming out `integration` from all those file names, which seems like mostly filler.)

File: ingest/ingest_pipeline.py [hunk 2]
```diff
@@ -519,44 +527,64 @@ def run_ingest(ingest, arguments, parsed_args):
...
+                    "log.txt",
+                    f"parse_logs/{study_file_id}/log.txt",
+                )
+                # Delocalize user log
+                IngestFiles.delocalize_file(
+                    study_file_id,
+                    arguments["study_id"],
+                    file_path,
+                    "user_log.txt",
+                    f"parse_logs/{study_file_id}/user_log.txt",
```
Comment (eweitz) [inline ingest/ingest_pipeline.py:586]: Putting DE log output into a DE-specific folder might help debuggability, by distinguishing DE / non-DE at a glance.

```suggestion
                    f"parse_logs/differential_expression/{study_file_id}/log.txt",
                )
                # Delocalize user log
                IngestFiles.delocalize_file(
                    study_file_id,
                    arguments["study_id"],
                    file_path,
                    "user_log.txt",
                    f"parse_logs/differential_expression/{study_file_id}/user_log.txt",
```

**Review by eweitz — CHANGES_REQUESTED** [review]
Solid progress!

I found a single functional blocker, regarding DE file path URLs.  I also suggest a handful of non-blocking refinements that are somewhat subjective, but maybe persuasive.

File: ingest/de.py [hunk 3]
```diff
@@ -364,3 +377,24 @@ def run_scanpy_de(
...
+        """
+
+        files = glob.glob(files_to_match)
+        for file in files:
+            IngestFiles.delocalize_file(
+                study_file_id,
+                None,
+                file_path,
+                file,
+                f"_scp_internal/differential-expression/{file}",
```
[the remark this answers is not in the document]
Comment (bistline) [reply ingest/de.py:399]: We don't reference any paths yet for differential expression outputs.  I imagine at some point when annotations get deleted, if they have DE outputs, we'll remove them from the bucket.  But as of now, we don't have them referenced.

Also, I agree with standardization on underscore.  Both Python and Ruby are "snake case", so it makes sense.

File: ingest/ingest_pipeline.py (hunk 1 above)
Comment (jlchang) [reply ingest/ingest_pipeline.py:39]: `b709b84` agreed. I've updated the files names.

File: ingest/ingest_pipeline.py (hunk 2 above)
Comment (jlchang) [reply ingest/ingest_pipeline.py]: `b709b84`

File: ingest/de.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jlchang) [reply ingest/de.py]: `b709b84`

**Review by eweitz — APPROVED** [review]

**Pull request merged**